lib/gossip/damper.js

```diff
@@ -0,0 +1,580 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var DampReqRequest = require('../../request_response.js').DampReqRequest;
+var EventEmitter = require('events').EventEmitter;
+var events = require('./events.js');
+var globalTimers = require('timers');
+var LoggingLevels = require('../logging/levels.js');
+var TypedError = require('error/typed');
+var util = require('util');
+
+var UnattainableRValError = TypedError({
+    type: 'ringpop.damping.unattainable-rval',
+    message: 'Unable to attain damp-req r-val',
+    rVal: null,
+    errors: null
+});
+
+function Damper(opts) {
+    this.ringpop = opts.ringpop;
+    this.timers = opts.timers || globalTimers;
+    this.Date = opts.Date || Date;
+    this.logger = this.ringpop.loggerFactory.getLogger('damping');
+
+    // Flappers are members with damping potential. They have
+    // exceeded the suppression limit and are considered as members
+    // that may need damping. They are no longer considered flappy
+    // when their damp score falls under the reuse limit.
+    this.flappers = {}; // indexed by address (id)
+
+    // TODO This is only temporary collection of damped members
+    // since we're not actually damping them (marking the
+    // member's status to damped) at this point. Damped members
+    // are removed from this collection when the expiration timer
+    // removes them.
+    this.dampedMembers = {}; // indexed by address
+
+    // This damp timer runs only when there are flappy members
+    // add to the flappers collection above.
+    this.isDampTimerEnabled = false;
+    this.dampTimer = null;
```

Comment:
Would dampTimer ever be truthy, but Enabled be false? If not, should isDampTimerEnabled be a function?
Same question for expiration timer below.